node: shut down all node-related HTTP servers gracefully

seh opened the pull request:
Rather than just closing the underlying network listener to stop our HTTP servers (including WebSocket), use [the graceful shutdown procedure](https://pkg.go.dev/net/http?tab=doc#Server.Shutdown), waiting for any in-process requests to finish.

Don't return the `net.Listener` to callers of `node.StartHTTPEndpoint`, as the `net/http.Server` assumes responsibility for the `net.Listener` lifecycle once fed to [`(*Server).Serve`](https://pkg.go.dev/net/http?tab=doc#Serve).

seh commented:
The `TestBoundedForkedSync63Fast` test doesn't look to be related to this change.

renaynay commented on node/node.go:
```diff
@@ -375,43 +379,42 @@ func (n *Node) startHTTP(endpoint string, apis []rpc.API, modules []string, cors
-		n.log.Info("HTTP endpoint closed", "url", url)
+	if n.httpServer != nil {
+		// Don't bother imposing a timeout here.
+		n.httpServer.Shutdown(context.Background())
```
Is there a reason for removing the log? 

I would still log that the server has been shut down.
```suggestion
		n.httpServer.Shutdown(context.Background())
		url := fmt.Sprintf("http://%v/", n.httpListenerAddr)
		n.log.Info("HTTP endpoint closed", "url", url)
```

seh replied:
No, no intentional reason. It was a symptom of the order in which I went about these changes. I added the calls to `(*http.Server).Shutdown`, but had left the `net.Listener.Close` calls in place. Once I realized that the latter were no longer necessary, I went through and removed them, but failed to preserve the log messages lost as collateral damage.

renaynay commented on node/node.go:
```diff
@@ -424,26 +427,25 @@ func (n *Node) startWS(endpoint string, apis []rpc.API, modules []string, wsOrig
-		n.log.Info("WebSocket endpoint closed", "url", fmt.Sprintf("ws://%s", n.wsEndpoint))
+	if n.wsHTTPServer != nil {
+		// Don't bother imposing a timeout here.
+		n.wsHTTPServer.Shutdown(context.Background())
```
Same here, is there a reason for removing the log? 

```suggestion
		n.wsHTTPServer.Shutdown(context.Background())
		url := fmt.Sprintf("ws://%v", n.wsListenerAddr)
		n.log.Info("WebSocket endpoint closed, "url", url)
```

renaynay commented on node/node.go:
```diff
@@ -424,26 +427,25 @@ func (n *Node) startWS(endpoint string, apis []rpc.API, modules []string, wsOrig
 		return err
 	}
-	n.log.Info("WebSocket endpoint opened", "url", fmt.Sprintf("ws://%s", listener.Addr()))
+	n.log.Info("WebSocket endpoint opened", "url", fmt.Sprintf("ws://%s", addr))
```
```suggestion
	n.log.Info("WebSocket endpoint opened", "url", fmt.Sprintf("ws://%v", addr))
```

seh commented:
Ouch. Something went wrong there. Let me rebase this change and push again.

seh commented:
I pushed a second commit (45e2022079b6d91f152586bb01cfec4a5b3048a3) addressing code review suggestions. We can squash that into the previous commit once we're ready to proceed.
